File: packages/honker-bun/test/python_interop.test.ts
import { test, expect } from "bun:test";
import { existsSync, mkdtempSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { delimiter, join, resolve } from "node:path";
import { spawnSync } from "node:child_process";

import { open } from "../src/index.ts";

const REPO_ROOT = resolve(import.meta.dir, "..", "..", "..");
const PYTHONPATH = [
  join(REPO_ROOT, "packages", "honker", "python"),
  join(REPO_ROOT, "packages"),
].join(delimiter);

function findPython(): string | null {
  const probeScript = `
import os
import tempfile
import honker

p = tempfile.mktemp(prefix="honker-probe-", suffix=".db")
db = honker.open(p)
db.query("SELECT 1")
db = None
try:
    os.remove(p)
except OSError:
    pass
`;
  const candidates = [
    process.env.HONKER_INTEROP_PYTHON,
    join(REPO_ROOT, ".venv", process.platform === "win32" ? "Scripts/python.exe" : "bin/python"),
    "python3",
    "python",
  ].filter(Boolean) as string[];

  for (const python of candidates) {
    const probe = spawnSync(python, ["-c", probeScript], {
      env: { ...process.env, PYTHONPATH },
      stdio: "ignore",
    });
    if (probe.status === 0) return python;
  }

  if (process.env.HONKER_INTEROP_PYTHON) {
    throw new Error("HONKER_INTEROP_PYTHON cannot import honker");
  }
  return null;
}

function findExtension(): string | null {
  const candidates = [
    process.env.HONKER_EXT_PATH,
    join(REPO_ROOT, "target/release/libhonker_ext.dylib"),
    join(REPO_ROOT, "target/release/libhonker_ext.so"),
  ].filter(Boolean) as string[];
  return candidates.find((p) => existsSync(p)) ?? null;
}

function runPython(python: string, dbPath: string, script: string): string {
  const out = spawnSync(python, ["-c", script], {
    env: { ...process.env, PYTHONPATH, DB_PATH: dbPath },
    encoding: "utf8",
  });
  expect(out.status, out.stderr || out.stdout).toBe(0);
  return out.stdout;
}

test("bun and python share queue, stream, and notify rows", () => {
  const python = findPython();
  const extPath = findExtension();
  if (!python || !extPath) {
    if (process.env.CI) throw new Error("python or honker extension missing in CI");
    return;
  }

  const dir = mkdtempSync(join(tmpdir(), "honker-bun-python-"));
  const dbPath = join(dir, "bun-python.db");
  const db = open(dbPath, extPath);
  try {
    const bunQueue = db.queue("bun-to-python");
    for (let i = 0; i < 25; i += 1) {
      bunQueue.enqueue({ source: "bun", seq: i, key: `bun-${i.toString().padStart(2, "0")}` });
    }
    db.notify("from-bun", { source: "bun", count: 25 });
    db.stream("interop").publish({ source: "bun", kind: "stream" });

    const observed = JSON.parse(runPython(python, dbPath, `
import json
import os
import honker

db = honker.open(os.environ["DB_PATH"])

jobs = db.queue("bun-to-python").claim_batch("python-worker", 50)
payloads = [job.payload for job in jobs]
acked = db.queue("bun-to-python").ack_batch([job.id for job in jobs], "python-worker")
note = db.query(
    "SELECT payload FROM _honker_notifications "
    "WHERE channel='from-bun' ORDER BY id DESC LIMIT 1"
)
events = db.stream("interop")._read_since(0, 10)

py_q = db.queue("python-to-bun")
for i in range(25):
    py_q.enqueue({"source": "python", "seq": i, "key": f"py-{i:02d}"})
with db.transaction() as tx:
    tx.notify("from-python", {"source": "python", "count": len(jobs)})
db.stream("interop").publish({"source": "python", "kind": "stream"})

print(json.dumps({
    "acked": acked,
    "payloads": payloads,
    "note": json.loads(note[0]["payload"]),
    "event_count": len(events),
}))
`));

    expect(observed.acked).toBe(25);
    expect(observed.payloads).toHaveLength(25);
    expect(observed.note).toEqual({ source: "bun", count: 25 });
    expect(observed.event_count).toBe(1);

    const pyQueue = db.queue("python-to-bun");
    const pyJobs = pyQueue.claimBatch("bun-worker", 50);
    expect(pyJobs).toHaveLength(25);
    const seqs = pyJobs.map((job) => (job.payload as any).seq).sort((a, b) => a - b);
    expect(seqs).toEqual(Array.from({ length: 25 }, (_, i) => i));
    expect(pyQueue.ackBatch(pyJobs.map((job) => job.id), "bun-worker")).toBe(25);

    const note = db.raw
      .query<{ payload: string }, []>(
        "SELECT payload FROM _honker_notifications WHERE channel='from-python' ORDER BY id DESC LIMIT 1",
      )
      .get()!;
    expect(JSON.parse(note.payload)).toEqual({ source: "python", count: 25 });
    expect(db.stream("interop").readSince(0, 10)).toHaveLength(2);
  } finally {
    db.close();
    rmSync(dir, { recursive: true, force: true });
  }
});

File: packages/honker-ex/test/python_interop_test.exs
defmodule HonkerPythonInteropTest do
  use ExUnit.Case, async: false

  alias Honker.{Job, Queue, Stream}

  @repo_root Path.expand("../../..", __DIR__)

  defp python_path do
    Enum.join(
      [
        Path.join([@repo_root, "packages", "honker", "python"]),
        Path.join([@repo_root, "packages"])
      ],
      if(:os.type() |> elem(0) == :win32, do: ";", else: ":")
    )
  end

  defp find_python do
    probe = """
    import os
    import tempfile
    import honker

    p = tempfile.mktemp(prefix="honker-probe-", suffix=".db")
    db = honker.open(p)
    db.query("SELECT 1")
    db = None
    try:
        os.remove(p)
    except OSError:
        pass
    """

    candidates =
      [
        System.get_env("HONKER_INTEROP_PYTHON"),
        Path.join([@repo_root, ".venv", "bin", "python"]),
        Path.join([@repo_root, ".venv", "Scripts", "python.exe"]),
        "python3",
        "python"
      ]
      |> Enum.reject(&is_nil/1)

    Enum.find(candidates, fn python ->
      {_, status} =
        System.cmd(python, ["-c", probe],
          env: [{"PYTHONPATH", python_path()}],
          stderr_to_stdout: true
        )

      status == 0
    end)
  end

  defp find_extension do
    [
      System.get_env("HONKER_EXTENSION_PATH"),
      System.get_env("HONKER_EXT_PATH"),
      Path.join([@repo_root, "target", "release", "libhonker_ext.dylib"]),
      Path.join([@repo_root, "target", "release", "libhonker_ext.so"])
    ]
    |> Enum.reject(&is_nil/1)
    |> Enum.find(&File.exists?/1)
  end

  setup do
    python = find_python()
    ext = find_extension()

    if (python == nil or ext == nil) and System.get_env("CI") do
      flunk("python honker binding or honker extension missing in CI")
    end

    if python == nil or ext == nil do
      {:ok, skip: true}
    else
      dir = Path.join(System.tmp_dir!(), "honker-ex-python-#{System.unique_integer([:positive])}")
      File.mkdir_p!(dir)
      db_path = Path.join(dir, "elixir-python.db")
      on_exit(fn -> File.rm_rf!(dir) end)
      {:ok, db} = Honker.open(db_path, extension_path: ext)
      {:ok, db: db, db_path: db_path, python: python}
    end
  end

  test "elixir and python share queue stream and notify rows", ctx do
    if Map.get(ctx, :skip) do
      :ok
    else
      for i <- 0..24 do
        {:ok, _} =
          Queue.enqueue(ctx.db, "elixir-to-python", %{
            "source" => "elixir",
            "seq" => i,
            "key" => "elixir-#{String.pad_leading(Integer.to_string(i), 2, "0")}"
          })
      end

      {:ok, _} = Honker.notify(ctx.db, "from-elixir", %{"source" => "elixir", "count" => 25})
      {:ok, _} = Stream.publish(ctx.db, "interop", %{"source" => "elixir", "kind" => "stream"})

      script = """
      import json
      import os
      import honker

      db = honker.open(os.environ["DB_PATH"])

      jobs = db.queue("elixir-to-python").claim_batch("python-worker", 50)
      payloads = [job.payload for job in jobs]
      acked = db.queue("elixir-to-python").ack_batch([job.id for job in jobs], "python-worker")
      note = db.query(
          "SELECT payload FROM _honker_notifications "
          "WHERE channel='from-elixir' ORDER BY id DESC LIMIT 1"
      )
      events = db.stream("interop")._read_since(0, 10)

      py_q = db.queue("python-to-elixir")
      for i in range(25):
          py_q.enqueue({"source": "python", "seq": i, "key": f"py-{i:02d}"})
      with db.transaction() as tx:
          tx.notify("from-python", {"source": "python", "count": len(jobs)})
      db.stream("interop").publish({"source": "python", "kind": "stream"})

      print(json.dumps({
          "acked": acked,
          "payloads": payloads,
          "note": json.loads(note[0]["payload"]),
          "event_count": len(events),
      }))
      """

      {out, status} =
        System.cmd(ctx.python, ["-c", script],
          env: [{"PYTHONPATH", python_path()}, {"DB_PATH", ctx.db_path}],
          stderr_to_stdout: true
        )

      assert status == 0, out
      observed = Jason.decode!(out)
      assert observed["acked"] == 25
      assert length(observed["payloads"]) == 25
      assert observed["note"] == %{"source" => "elixir", "count" => 25}
      assert observed["event_count"] == 1

      {:ok, py_jobs} = Queue.claim_batch(ctx.db, "python-to-elixir", "elixir-worker", 50)
      assert length(py_jobs) == 25
      assert Enum.map(py_jobs, & &1.payload["seq"]) |> Enum.sort() == Enum.to_list(0..24)
      assert Enum.all?(py_jobs, &(&1.payload["source"] == "python"))
      assert Enum.all?(py_jobs, fn job -> Job.ack(ctx.db, job) == {:ok, true} end)

      {:ok, [note]} =
        Honker.query_first(
          ctx.db.conn,
          "SELECT payload FROM _honker_notifications WHERE channel='from-python' ORDER BY id DESC LIMIT 1",
          []
        )

      assert Jason.decode!(note) == %{"source" => "python", "count" => 25}
      {:ok, events} = Stream.read_since(ctx.db, "interop", 0, 10)
      assert length(events) == 2
    end
  end
end
